Add PInvoke/Structures tests

jiangzeng opened the pull request.

jiangzeng01 committed:
Add PInvoke/Structures tests

Fix compile warnings

Fix compile warnings

Fix compile warnings

Remove duplicate tests. Move ReversePinvoke tests under StructMarshalling.

Clean up some compile errors.

More string cleanup.

More string fixes

Clean up consts and typedefs.

Fix typo.

Change TP_slen return type to match xplatform.h

Fix printf specifiers.

Don't define wcslen in xplatform if it's been macro redefined (to platformdefines.h's TP_slen)

Fix build breaks.

Remove extension from DllImport so it will correctly resolve xplat.

Fix printf specifiers again...

Fix Windows x86 calling convention mismatch.

Fix HRESULT size off-Windows

jiangzeng commented on tests/src/Interop/PInvoke/Structures/PInvoke/AsParamNative.h:
```diff
@@ -0,0 +1,954 @@
+void ChangeStringExplicit(StringExplicit* p)
+{
+    LPCSTR str1 = "Managed_LPCSTR\0";
+    LPCTSTR str2 = (LPCTSTR)L"Managed_LPCSTR\0";
```
@AaronRobinsonMSFT @luqunl There's warning here, do you have any idea about how to fix?
`warning C4905: wide string literal cast to 'LPCTSTR'`

AaronRobinsonMSFT replied:
Boooo. Get rid of that type. Please convert all instances of `LPCTSTR` into `LPCWSTR`. The former is the TCHAR logic that always confuses people.

jiangzeng replied:
@AaronRobinsonMSFT Thanks for your suggestion, this works

jiangzeng commented on tests/src/Interop/PInvoke/Structures/PInvoke/AsParamNative.cpp:
```diff
@@ -0,0 +1,1229 @@
+        return FALSE;
+    else
+    {
+        str1->i32 = (LPINT)(str1->i);
```
@AaronRobinsonMSFT @luqunl There's warning here, do you have any idea about how to fix?
`warning C4312: 'type cast': conversion from 'INT' to 'LPINT' of greater size`

AaronRobinsonMSFT replied:
This is similar to the `size_t` issue. The problem here is that `i32` is a 32-bit int, but we are trying to assign a pointer - which on 64-bit build is 64-bit - hence the warning.

AaronRobinsonMSFT replied:
Try first casting to a `UINT`

jiangzeng replied:
@AaronRobinsonMSFT Thanks for your suggestion, I tried UINT but failed with similar issue.
So I tried INT64 as first cast like `str1->i32 = (LPINT)(INT64)(str1->i)`, and it works without warnings, do you think INT64 here is reasonable?

AaronRobinsonMSFT commented on tests/src/Interop/PInvoke/Structures/PInvoke/AsParamNative.h:
```diff
@@ -0,0 +1,954 @@
+    printf("\t%s.i32 = %d\n", name, str->i32);
+    printf("\t%s.ui32 = %u\n", name, str->ui32);
+    printf("\t%s.iPtr = %d\n", name, (size_t)str->iPtr);
+    printf("\t%s.uiPtr = %u\n", name, (size_t)str->uiPtr);
```
This should only an issue on 64-bit when `size_t` is 64-bit. The issue here is that `size_t` can have different sizes depending on build. For Microsoft compilers you could use the `%Iu` formatter, but that won't work on non-Microsoft compilers. Instead I would try `%zu`.

jiangzeng replied:
@AaronRobinsonMSFT Thanks for your suggestion, %zu works without warnings

jiangzeng commented on tests/src/Interop/PInvoke/Structures/DelegatePInvoke/MarshalExpStruct/ReversePInvokeManaged/ReversePInvokeTest.cs:
```diff
@@ -0,0 +1,807 @@
+            DoCallBack_MarshalByValStruct_Cdecl_LongStructPack16Explicit(new ByValCdeclcaller_LongStructPack16Explicit(TestMethod_DoCallBack_MarshalStructByVal_LongStructPack16Explicit_Cdecl));
+
+            ////Reverse Pinvoke,ByVal,stdcall
+            DoCallBack_MarshalByValStruct_Stdcall_INNER2(new ByValStdcallcaller_INNER2(TestMethod_DoCallBack_MarshalStructByVal_INNER2_Stdcall));
```
This test scenario failed in X86 but pass in X64 do you have any idea?
`Test Failure: System.NullReferenceException: Object reference not set to an instance of an object.
20:07:09            at Test.DoCallBack_MarshalByValStruct_Stdcall_INNER2(ByValStdcallcaller_INNER2 caller)
20:07:09            at Test.Main() in D:\j\workspace\x86_checked_w---7329ffdd\tests\src\Interop\PInvoke\Structures\DelegatePInvoke\MarshalExpStruct\ReversePInvokeManaged\ReversePInvokeTest.cs:line 787`

AaronRobinsonMSFT commented on tests/src/Interop/PInvoke/Structures/DelegatePInvoke/MarshalSeqStruct/SeqStructDelRevPInvokeNative.cpp:
```diff
@@ -0,0 +1,3240 @@
+    //Init
+    InnerSequential argstr;
+    argstr.f1 = 77;
+    argstr.f2 = 77.0;
```
`InnerSequential argstr{};`?

jiangzeng replied:
Thanks for your suggestion, it works

jiangzeng commented on tests/src/Interop/PInvoke/Structures/DelegatePInvoke/MarshalSeqStruct/SeqStructDelRevPInvokeNative.cpp:
```diff
@@ -0,0 +1,3240 @@
+extern "C" DLL_EXPORT BOOL _cdecl MarshalStructComplexStructByRef_Cdecl(ComplexStruct * pcs)
+{                       
+    //Check the Input    
+    if((321 != pcs->i)||(!pcs->b)||(0 != strcmp(pcs->str,"Managed"))||(123 != pcs->type.idata)||(0x120000 != reinterpret_cast<int>(pcs->type.udata.ptrdata)))
```
Thanks. Fixed it with `(int)(INT64)(pcs->type.udata.ptrdata)`

jiangzeng commented on tests/src/Interop/PInvoke/Structures/DelegatePInvoke/MarshalSeqStruct/SeqStructDelRevPInvokeNative.h:
```diff
@@ -0,0 +1,895 @@
+        LONG64 _unused0;
+        LPCSTR f4;
+    }; 
+
```
I'm sorry, this code is never used, so I deleted it.
Please see the following comment related to this.

jiangzeng commented on tests/src/Interop/PInvoke/Structures/DelegatePInvoke/MarshalSeqStruct/SeqStructDelRevPInvokeNative.h:
```diff
@@ -0,0 +1,884 @@
+    {
+        CHAR _unused0[24];
+        LPCSTR f4;
+    };
```
Warning about unnamed struct here.
`warning C4201: nonstandard extension used: nameless struct/union`
I tried to add Inner to this struct, issue would happen, f4 will be unreachable by code example `p->f4`.
This will cause issue in code. As if the struct is named, we cannot access f4 by this way.

jiangzeng replied:
Is it ok to just define OUTER3 as following but not using `union` in X64:
`struct OUTER3 // size = 28 bytes
{
    struct InnerSequential arr[2];
    LPCSTR f4;
};`

AaronRobinsonMSFT replied:
I don't understand the issue then. Why can't we fix access of that field in code?

jiangzeng replied:
@AaronRobinsonMSFT The definition for OUTER3 is different in X64 & X86 build according to the code.
If we added Inner to the unnamed struct in X64, then the access to the f4 field will be different in X86 & X64.
X64: `p.Inner.f4`
X86:`p.f4`

If I changed the X86 struct OUTER similiar to X64, added Inner struct, then move f4 into Inner struct. 
`struct OUTER3 // size = 28 bytes
{
    struct InnerSequential arr[2];
    struct Inner
    {
        LPCSTR f4;
    }
};`
Build passed, but the C# test failed with BadImageFormatException. 
And I'm not sure why OUTER3 need to be different in X86 & X64.
So I'm not sure how to fix the C# test.

janvorli commented:
@jkoritzinsky the RSP cannot point to the place where the original stack arguments are located, since we have the `UMThunkStub`'s frame below that. The original stack arguments should get copied by the `UMThunkStub` to a new location. So from what you say it seems it doesn't happen for some reason. 
The `UMThunkStubs` gets the number of stack arguments here:
https://github.com/dotnet/coreclr/blob/master/src/vm/amd64/umthunkstub.S#L117
and if that's not zero, it copies them.
The variable is `UMThunkMarshInfo::m_cbActualArgSize`. It is computed in `UMThunkMarshInfo::CompileNExportThunk` or `UMThunkMarshInfo::RunTimeInit`.
If you have environment setup for native code debugging on Linux, you can put a breakpoint into the `UMThunkStub` and see what size it gets. And if it is wrong, try to debug one of the `UMThunkMarshInfo` methods that generate it.
If not, I can help you to debug that.

jkoritzinsky commented:
When it's computed in `UMThunkMarshInfo::RunTimeInit` it is initially computed correctly to be 8 at https://github.com/dotnet/coreclr/blob/master/src/vm/dllimport.cpp#L4077 but is then updated to 0 at 
https://github.com/dotnet/coreclr/blob/master/src/vm/dllimport.cpp#L607 during the `FinishEmit` call in `SwapStubSignatures`.

jkoritzinsky commented:
As far as I can tell, the incorrectly calculated value came from 
https://github.com/dotnet/coreclr/blob/master/src/vm/callingconvention.h#L1510-L1672

jkoritzinsky commented:
Ok so the correct value for `m_nSizeOfArgStack` in `ArgIteratorTemplate<ArgIteratorBase>::ForceSigWalk` should be 32. However, since the argument is mistakenly thought to be in a register, the value of `m_nSizeOfArgStack` is set to 0.

jkoritzinsky commented:
The method table for the struct is having it's `enum_flag_IsRegStructPassed` flag set to true, so the `ArgIteratorTemplate<ArgIteratorBase>` class thinks that the value is being passed in registers instead of on the stack. If I manually make `MethodTable::IsRegPassedStruct` return `false`, then everything works.

@janvorli any ideas why that's happening? I'm going to try to debug it, but I can already tell that this one is going to be much harder to trace than the previous steps I've done.

janvorli commented:
The flag is set based on the result of a call to MethodTable::ClassifyEightBytes. This function uses SystemV amd64 ABI rules to classify the struct to see how it should be passed. My guess is that the `[MarshalAs(UnmanagedType.ByValArray, SizeConst =2)]` is somehow not having effect.

jkoritzinsky commented:
@janvorli I think I found the reason that `ArgIteratorTemplate<ArgIteratorBase>` is wrong and the jitted ILStub is correct.

The managed structure is calculated to be passed via registers. In `ArgIteratorTemplate<ArgIteratorBase>::GetNextOffset()`, the algorithm only checks if the managed struct is passed via registers, which it is.

However, the native struct is not passed in registers, which the SystemV classifier correctly calculates and sets in the class layout info (`pMT->GetLayoutInfo()->IsNativeStructPassedInRegisters()` returns `false`). I think the JIT correctly checks the native struct instead of the managed one, so it knows to pull the parameter from the stack. However, since `ArgIteratorTemplate<ArgIteratorBase>::GetNextOffset()` thinks that it's passed via registers, the `m_nSizeOfArgStack` is set to `0`, which leads to nothing being copied from the stack.

I could change this part of the code to check the native structure instead of the managed one (especially since the argument size is already set to the correct native size of 32 for some reason I haven't investigated yet), but I'm not sure if that's the correct fix, or if it will fix it at all. What's your suggestion?

jkoritzinsky commented:
Ok so apparently if `TypeHandle` represents a value type and is a `TypeDesc` then `TypeHandle::GetSize` returns the native size of the struct.

janvorli commented:
@jkoritzinsky this ArgIteratorTemplate is used for managed structures passing during reflection, so it seems we would likely need to add some flag that would differentiate between the managed and native cases.

jkoritzinsky commented:
I'll file a separate bug for this so we can track it more easily.

janvorli commented:
Or maybe just the different template parameter that we use for reflection and pinvoke can be the differentiating thing. I have not seen this code for quite some time, but I used to be very familiar with it during the Unix bringup. 
I can look into fixing it unless you are interested in doing that.

jkoritzinsky commented:
Is that what the other template parameter value is used for? If that's the differentiator then I can probably try to fix this pretty easily.

janvorli commented:
@jkoritzinsky yes, there are two instantiations of that iterator:
`class ArgIterator : public ArgIteratorTemplate<ArgIteratorBase>` - this one is used for pinvoke

`class ArgIteratorForMethodInvoke : public ArgIteratorTemplate<ArgIteratorBaseForMethodInvoke>` - this one is used for reflection

jkoritzinsky commented:
Perfect! I'll take a stab at fixing it then!

janvorli commented:
Hmm, actually, the `ArgIterator` is used for more than just pinvoke. So we may need to create a new one for pinvoke. That should not be difficult as the template parameter class seems to be very think.

jkoritzinsky commented:
Ill do that then.

jkoritzinsky committed:
Make the check for a struct fitting in a register check for either managed or native depending on the situation (i.e. managed call, reflection, P/Invoke) instead of only for managed. Fixes #20702.

Clean up duplicate #ifdefs.

Move #ifdef into method impl instead of outside the implementations.

Move IsRegPassedStruct out of UNIX_AMD64_ABI #ifdef.

Move check for enregistered struct out of UNIX_AMD64_ABI #ifdef

Add dummy implementation of IsRegPassedStruct and IsNativeStructPassedInRegisters when UNIX_AMD64_ABI isn't defined.

jkoritzinsky commented:
@janvorli I just pushed a fix out. Any chance you can make a review pass on the fix when you have a chance?

janvorli reviewed (commented):
The fix for check of passing structs in regs LGTM. I have not reviewed the new tests.

AaronRobinsonMSFT commented on src/vm/reflectioninvocation.cpp:
```diff
@@ -760,6 +760,15 @@ class ArgIteratorBaseForMethodInvoke
         LIMITED_METHOD_CONTRACT;
     }
 
+#ifdef UNIX_AMD64_ABI
```
I don't see the need for the `UNIX_AMD_ABI` since this is the default case for use.

jkoritzinsky replied:
The `MethodDesc::IsRegPassedStruct` (which this uses) is `#ifdef`'d in for only `UNIX_AMD64_ABI`. Additionally, this logic is only used when `UNIX_AMD64_ABI` is defined anyway (since that's the only time that it is relevant.

AaronRobinsonMSFT replied:
>Additionally, this logic is only used when UNIX_AMD64_ABI is defined anyway (since that's the only time that it is relevant.

Usage is fine to be `#ifdef` out, but definitions largely have no reason to be defined out _unless_ they won't compile or will lie and in this case neither should be an issue. They will compile just fine and will return the truth. I would prefer to use `#ifdef`s sparingly.

AaronRobinsonMSFT commented on src/vm/callingconvention.h:
```diff
@@ -1701,6 +1701,15 @@ class ArgIteratorBase
         m_pSig->Reset();
     }
 
+#ifdef UNIX_AMD64_ABI
```
I would remove this `#ifdef` as well.

AaronRobinsonMSFT commented on src/vm/method.cpp:
```diff
@@ -1753,6 +1774,24 @@ UINT MethodDesc::SizeOfArgStack()
     return argit.SizeOfArgStack();
 }
 
+#ifndef UNIX_AMD64_ABI
```
Move the `#ifdef` inside a single definition of the function.

jkoritzinsky commented on src/vm/callingconvention.h:
```diff
@@ -1017,7 +1017,7 @@ int ArgIteratorTemplate<ARGITERATOR_BASE>::GetNextOffset()
 #ifdef UNIX_AMD64_ABI
         MethodTable *pMT = m_argTypeHandle.GetMethodTable();
-        if (pMT->IsRegPassedStruct())
+        if (this->IsRegPassedStruct(pMT))
```
The `this` is needed for 2-phase lookup IIRC. In any case, it won't build on Clang without it (can't find the member). It was an MSVC extension that allowed it to be omitted.

janvorli replied:
Yes, clang is more strict here and requires it.

AaronRobinsonMSFT replied:
Bah. I wish we could all agree... Thanks for the reminder.

>It was an MSVC extension that allowed it to be omitted.

This isn't an MSVC extension - all extensions should be disabled. GCC will also probably accept this. As @janvorli stated, the issue here is the strict adherence to the 2 phase lookup algorithm implementation in clang.

jkoritzinsky replied:
MSVC didn't actually implement 2-phase lookup correctly until very recently. They had to re-write their frontend parsing pipeline to be able to do templates in a standards compliant manner. The standards-compliant behavior (which is what Clang has) is by default off, but can be turned on with the `/permissive-` switch.

AaronRobinsonMSFT replied:
@jkoritzinsky My point was this has nothing to do with MSVC extensions. They are different things and should be off by default.

AaronRobinsonMSFT reviewed (approved):
Thank you. LGTM.
